jlm/llvm/ir/operators/lambda.hpp

```diff
@@ -273,16 +259,27 @@ class node final : public jlm::rvsdg::structural_node
   size_t
   PruneLambdaInputs()
   {
-    auto match = [](const cvinput &)
+    auto match = [](const rvsdg::input &)
     {
       return true;
     };
 
     return RemoveLambdaInputsWhere(match);
   }
 
-  [[nodiscard]] cvinput *
-  input(size_t n) const noexcept;
+  /**
+   * @brief Get argument in region corresponding to input
+   *
+   * @param input Input to the lambda node
+   * @return The region argument corresponding to the input
+   *
+   * The given input must be an input to this lambda node.
+   * All inputs represent by definition bound context variables
+   * and map to corresponding arguments in the lambda defining
+   * region.
+   */
+  [[nodiscard]] rvsdg::RegionArgument *
+  MapInputToRegionArgument(const rvsdg::input * input) const noexcept;
```

Comment:
I find the name a mouthful. How about just `RegionArgument` or `GetRegionArgument`? I also think that the signature should be:
```
[[nodiscard]] rvsdg::RegionArgument *
RegionArgument(const rvsdg::input & input) const noexcept
```

Reply:
We could also make the return type `rvsdg::RegionArgument &` and just leave it to the programmer to ensure that a correct  input is handed in. An assert in the body of the method would then be appropriate though.

Reply:
Can change it, although I am less fond of references compared to pointers TBH. The semantic difference that references convey is that they tell "cannot be null". On the downside, they tend to cause trouble with accidental copies (`auto x = foo.GetSomething()` ), cannot directly go into `std::optional`/`std::variant`/containers  without sprinkling `&`/`*` adornments etc. I think that ideally we would have "safe wrapper types" with value semantics that act as input / output / node handles etc., cannot be null but behave sanely for copies etc. However I would not go there right now.

Reply:
> I find the name a mouthful. How about just `RegionArgument` or `GetRegionArgument`?

Yes the name is horrible. This is difficult to find name that is neither ambiguous nor too long. `RegionArgument` is already the name of the class, I would like to avoid clashing with that. Other ideas that I came up with:
- `InputToInner` or `ToInnerArgument` or `ToInner` as it maps an (outer) input to an (inner) argument (or arguments, for gamma then). `ToInner` would be less great as we need the equivalent for outputs and would then have an overloaded name.
- `ArgumentOf` or `ArgumentOfInput` or `RegionArgumentOf` or similar
- `ProjectInputToMatchingArgumentOfSubregion` would be the most precise

Reply:
The inputs should(!) be marked as explicitly deleted, so the accidental copies should never happen.

How about `MapToRegionArgument` or `MapToArgument`? The argument and return type convey what we map to what, respectively. I think this should give us some wiggle room in terms of naming.

Reply:
`MapToArgument` sounds good in principle -- I think we ought to have somewhat coherent API names for the following mapping operations:

- input of structural node to region argument(s) (applies to lambda, gamma, theta, phi, delta)
- output of structural node to region result(s) (applies to gamma, theta, phi, delta)
- and corresponding reverse mappings

So it would then be `MapToArgument` / `MapToArguments`, `MapToResult` / `MapToResults`, `MapToInput`, `MapToOutput` ? BTW this is where I thought the notion of inner / outer would explain the API quite nicely.

pointers vs. references: presently trying hard to have references, but it causes enormous tying to specific data structures -- e.g. for returning multiple arguments, I would be inclined to simply use `std::vector<RegionArgument*>`  for now at least, until the eventual structures become clear. If I have to maintain various custom iterator ranges for every use case that makes changes a bit slower, but I cannot do anything involving references with standard containers.

Also, we may have cases where there really isn't a mapping -- e.g. the theta continuation predicate does not go outwards. If I have a function mapping region results to outputs of the theta node I either have to throw, or have a way to express `std::optional` of the thing (which also would indeed be expressible by nullptr). I know that there is code that cares about semantically mapping region entry/exits to inputs/outputs, but there is also code just trying to generically trace through and would like to just see the continuation paths.

Reply:
> So it would then be MapToArgument / MapToArguments, MapToResult / MapToResults, MapToInput, MapToOutput ? BTW this is where I thought the notion of inner / outer would explain the API quite nicely.

The names you suggest sound fine to me. What I do not like about the `Inner`/`Outer` is that there is no counterpart. In other words, `ToInnerArgument` sounds off to me as there is no `ToOuterArgument`. Thus, the `Inner` is redundant here. A simple `ToArgument` or `MapToArgument` is less confusing in my opinion (the inner is implied as the method takes an input and returns an argument).

> pointers vs. references: presently trying hard to have references, but it causes enormous tying to specific data structures -- e.g. for returning multiple arguments, I would be inclined to simply use std::vector<RegionArgument*> for now at least, until the eventual structures become clear. If I have to maintain various custom iterator ranges for every use case that makes changes a bit slower, but I cannot do anything involving references with standard containers.

We currently (try to) use references (where applicable (i.e. no chance of copying) and it is always guaranteed to have a valid element) for individual instances (RegionArgument, RegionResult, etc.), but when we give back collections/containers of instances, then we currently use pointers, i.e. `std::vector<RegionArgument*>` is perfectly fine. I have not dig into how to use references in containers. As pointed out by you or @haved in another PR, it is possible, but it is fine without for now.

> Also, we may have cases where there really isn't a mapping -- e.g. the theta continuation predicate does not go outwards. If I have a function mapping region results to outputs of the theta node I either have to throw, or have a way to express std::optional of the thing (which also would indeed be expressible by nullptr). I know that there is code that cares about semantically mapping region entry/exits to inputs/outputs, but there is also code just trying to generically trace through and would like to just see the continuation paths.

If there is no guarantee that it is possible to get a valid element, then using pointers is perfectly fine (I think). The alternative is throwing an exception (not big fan of it) or leaving it to the user of the method, documenting it and putting asserts in the method to double check. That depends a little bit on the context I think.

Reply:
Used references in all places possible. Now that the overall API becomes clearer I have resorted to pointers in places necessary.

Looking at where mapping is done, it is clear that this 95% at construction time, and very little at tracing. So I think the best thing to do is to provide an API that returns a "description" of the compound thing being created. For reference, I also uploaded changes for gamma and theta to illustrate the concept, see https://github.com/phate/jlm/commit/217e8798714cbd628a6bff9c67f4f157881e7576 and https://github.com/phate/jlm/commit/63e38024acd6820540d549056449ab73ba90aaf0 -- particular example: https://github.com/phate/jlm/blob/217e8798714cbd628a6bff9c67f4f157881e7576/jlm/rvsdg/gamma.hpp#L230

So, caller gets what they need to proceed from call to AddEntryVar (or AddLoopVar, or AddContextVar). Also, all "mapping" operations that want to resolve the full EntryVar/LoopVar/ContextVar just supply the piece they "have" and can get the full thing resolved. I think this makes for a consistent and well-usable API. WDYT?